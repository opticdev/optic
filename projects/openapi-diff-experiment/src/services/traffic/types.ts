import events from 'events';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';
/*

  Interface for traffic, expose what the differ cares about via interfaces
   - different formats can have different implementations
   - debug one (just to get rolling)
     - HTTP Interaction
     - HAR
     - Postman?

   */
export interface ApiTraffic {
  method: OpenAPIV3.HttpMethods;
  path: string;
  queryString: string;
  response: {
    statusCode: string;
    body: {
      contentType?: string;
      jsonBodyString?: string;
    };
  };
  requestBody?: {
    contentType?: string;
    jsonBodyString?: string;
  };
}

/*

  Interface for traffic, should get traffic (somehow), and emit each interaction into our queue
   - each implementation should be self contained starting / parsing / stopping traffic collection
   - known types
     - file
     - in-memory (for debugging)
     - proxy
     - postman
     - sniff
     - chrome

   */

export interface ITrafficSource {
  start(): Promise<void>;
  stop(): Promise<void>;
  emitTraffic(traffic: ApiTraffic): Promise<void>;
  on(event: 'traffic', listener: (example: ApiTraffic) => void): this;
}

export class TrafficSource
  extends events.EventEmitter
  implements ITrafficSource
{
  emitTraffic(traffic: ApiTraffic): Promise<void> {
    this.emit('traffic', traffic);
    return Promise.resolve();
  }

  start(): Promise<void> {
    return Promise.resolve(undefined);
  }

  stop(): Promise<void> {
    return Promise.resolve(undefined);
  }
}
