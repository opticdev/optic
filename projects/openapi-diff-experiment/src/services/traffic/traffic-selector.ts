import { ApiTraffic } from './types';
import { ConceptualLocation } from '@useoptic/openapi-utilities';
import invariant from 'ts-invariant';

export function trafficSelector(traffic: ApiTraffic) {
  return {
    bodyJsonForLocation: (location: ConceptualLocation) => {
      if ('inResponse' in location && 'body' in location.inResponse) {
        const contentType = location.inResponse.body.contentType;

        invariant(
          contentType === traffic.response.body.contentType,
          'content type must match when selecting examples from traffic'
        );

        invariant(
          typeof traffic.response.body.jsonBodyString !== 'undefined',
          'body json must be present if selecting example from traffic'
        );

        return JSON.parse(traffic.response.body.jsonBodyString!);
      }

      if ('inRequest' in location && 'body' in location.inRequest) {
        const { contentType } = location.inRequest.body;
        invariant(
          contentType === traffic.requestBody?.contentType,
          'content type must match when selecting examples from traffic'
        );
        invariant(
          typeof traffic.requestBody.jsonBodyString !== 'undefined',
          'body json must be present if selecting example from traffic'
        );
        return JSON.parse(traffic.requestBody.jsonBodyString!);
      }

      invariant(false, 'can not select example body from traffic');
    },
  };
}
