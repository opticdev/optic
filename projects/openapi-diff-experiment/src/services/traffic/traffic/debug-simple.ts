import { ApiTraffic } from '../types';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';

export class SimpleExampleTraffic implements ApiTraffic {
  constructor(
    private _request: { method: string; path: string; queryString?: string },
    private _response: { statusCode: string; json?: any }
  ) {}
  get method() {
    return this._request.method.toLowerCase() as OpenAPIV3.HttpMethods;
  }
  get path() {
    return this._request.path;
  }
  get response() {
    return {
      statusCode: this._response.statusCode,
      body: {
        contentType: this._response.json ? 'application/json' : undefined,
        jsonBodyString: this._response.json
          ? JSON.stringify(this._response.json)
          : undefined,
      },
    };
  }

  get queryString() {
    return this._request.queryString || '';
  }
}

export function makeExample(
  path: string,
  method: string | OpenAPIV3.HttpMethods,
  statusCode: string = '200',
  jsonBody?: object | undefined
) {
  return new SimpleExampleTraffic(
    { path, method },
    { statusCode, json: jsonBody }
  );
}
