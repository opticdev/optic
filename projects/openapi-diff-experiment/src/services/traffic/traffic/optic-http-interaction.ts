import { ApiTraffic } from '../types';
import { OpenAPIV3 } from '@useoptic/openapi-utilities';

// TODO: Fill in the type
// {
//   "uuid": "c11116b1-cea5-401c-b151-da47ab43cee3",
//   "request": {
//     "host": "httpbin.org",
//     "method": "GET",
//     "path": "/get",
//     "headers": {
//       "shapeHashV1Base64": null,
//       "asJsonString": null,
//       "asText": null
//     },
//     "query": {
//       "shapeHashV1Base64": null,
//       "asJsonString": null,
//       "asText": "a=1"
//     },
//     "body": {
//       "contentType": null,
//       "value": {
//         "shapeHashV1Base64": null,
//         "asJsonString": null,
//         "asText": null
//       }
//     }
//   },
//   "response": {
//     "statusCode": 200,
//     "headers": {
//       "shapeHashV1Base64": null,
//       "asJsonString": null,
//       "asText": null
//     },
//     "body": {
//       "contentType": "application/json",
//       "value": {
//         "shapeHashV1Base64": "CAASEwoEYXJncxILCAASBwoBYRICCAISUAoHaGVhZGVycxJFCAASDAoGQWNjZXB0EgIIAhIKCgRIb3N0EgIIAhIQCgpVc2VyLUFnZW50EgIIAhIVCg9YLUFtem4tVHJhY2UtSWQSAggCEgwKBm9yaWdpbhICCAISCQoDdXJsEgIIAg==",
//         "asJsonString": "{\"args\":{\"a\":\"1\"},\"headers\":{\"Accept\":\"*/*\",\"Host\":\"httpbin.org\",\"User-Agent\":\"curl/7.64.1\",\"X-Amzn-Trace-Id\":\"Root=1-618d1bfc-0c9eea1060c29d6348b62f19\"},\"origin\":\"31.10.148.195\",\"url\":\"http://httpbin.org/get?a=1\"}",
//         "asText": null
//       }
//     }
//   },
//   "tags": []
// }
type IOpticCaptureHttpInteraction = any;

export class OpticHttpInteraction implements ApiTraffic {
  private _method: OpenAPIV3.HttpMethods;
  private _path: string;
  private _queryString: string;
  private _requestBody: ApiTraffic['requestBody'];
  private _response: {
    statusCode: string;
    body: {
      contentType: string | undefined;
      jsonBodyString: string | undefined;
    };
  };

  constructor(_interaction: IOpticCaptureHttpInteraction) {
    this._method =
      _interaction.request.method.toLowerCase() as OpenAPIV3.HttpMethods;
    this._path = _interaction.request.path;
    this._queryString = _interaction.request.query.asText || '';
    this._requestBody = _interaction.request.body && _interaction.request.body.contentType
      ? {
          contentType: _interaction.request.body.contentType,
          jsonBodyString:
            _interaction.request.body.value.asJsonString || undefined,
        }
      : undefined;
    this._response = {
      statusCode: _interaction.response.statusCode.toString(),
      body: {
        contentType: _interaction.response.body && _interaction.response.body.contentType
          ? 'application/json'
          : undefined,
        jsonBodyString: _interaction.response.body.value.asJsonString
          ? _interaction.response.body.value.asJsonString
          : undefined,
      },
    };
  }

  get method() {
    return this._method;
  }
  get path() {
    return this._path;
  }
  get requestBody() {
    return this._requestBody;
  }
  get response() {
    return this._response;
  }

  get queryString() {
    return this._queryString;
  }
}
