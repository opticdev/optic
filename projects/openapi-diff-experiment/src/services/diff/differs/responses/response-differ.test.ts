import {
  openApiDiffingQuestionsTestingStub,
  ResponseMatchType,
} from '../../../read/types';
import {
  BodyLocation,
  ConceptualLocation,
  OpenAPIV3,
} from '@useoptic/openapi-utilities';
import { responsesDiffer } from './index';
import { makeExample } from '../../../traffic/traffic/debug-simple';

const noResponsesFixture = responsesDiffer({
  ...openApiDiffingQuestionsTestingStub,
  responsesForOperation(
    method: OpenAPIV3.HttpMethods,
    path: string
  ): ResponseMatchType[] {
    return [];
  },
});

const someResponsesFixture = responsesDiffer({
  ...openApiDiffingQuestionsTestingStub,
  responsesForOperation(
    method: OpenAPIV3.HttpMethods,
    path: string
  ): ResponseMatchType[] {
    return [
      {
        statusCodeMatcher: '200',
        contentTypes: [],
      },
      { statusCodeMatcher: '4xx', contentTypes: [] },
      { statusCodeMatcher: 'default', contentTypes: [] },
    ];
  },
});

const exampleWithStatusCode = (statusCode: string) =>
  makeExample('/example', OpenAPIV3.HttpMethods.GET, statusCode);

const exampleMatch = {
  pathParameterValues: {},
  method: OpenAPIV3.HttpMethods.GET,
  path: '/example',
  urlPath: '/example',
};

const mockLocation = (
  statusCode: string,
  contentType: string
): BodyLocation => ({
  path: '/example',
  method: OpenAPIV3.HttpMethods.GET,
  inResponse: {
    statusCode,
    body: {
      contentType,
    },
  },
});

describe('response differ', () => {
  it('matches a known response by exact status code', () => {
    expect(
      someResponsesFixture.responseDiffsForTraffic(
        exampleWithStatusCode('200'),
        exampleMatch
      )
    ).toMatchSnapshot();
  });

  it('matches a known response by range of status code', () => {
    expect(
      someResponsesFixture.responseDiffsForTraffic(
        exampleWithStatusCode('404'),
        exampleMatch
      )
    ).toMatchSnapshot();
  });

  it('matches default response status code if not other matches', () => {
    expect(
      someResponsesFixture.responseDiffsForTraffic(
        exampleWithStatusCode('201'),
        exampleMatch
      )
    ).toMatchSnapshot();
  });

  it('diff for new status code', () => {
    expect(
      noResponsesFixture.responseDiffsForTraffic(
        exampleWithStatusCode('200'),
        exampleMatch
      )
    ).toMatchSnapshot();
  });

  it('no diff for 500 range status code', () => {
    expect(
      noResponsesFixture.responseDiffsForTraffic(
        exampleWithStatusCode('500'),
        exampleMatch
      )
    ).toMatchSnapshot();
  });

  const someResponsesWithBodiesFixture = responsesDiffer({
    ...openApiDiffingQuestionsTestingStub,
    responsesForOperation(
      method: OpenAPIV3.HttpMethods,
      path: string
    ): ResponseMatchType[] {
      return [
        {
          statusCodeMatcher: '200',
          contentTypes: [
            {
              jsonPath: '',
              contentType: 'application/json',
              schema: {
                type: 'object',
                properties: { one: { type: 'string' } },
              },
              location: mockLocation('200', 'application/json'),
            },
          ],
        },
        {
          statusCodeMatcher: '4xx',
          contentTypes: [
            {
              jsonPath: '',
              contentType: 'text/plain',
              location: mockLocation('4xx', 'text/plain'),
            },
          ],
        },
        { statusCodeMatcher: 'default', contentTypes: [] },
      ];
    },
  });

  it('json schema diffs found in responses', () => {
    const traffic = makeExample('/example', OpenAPIV3.HttpMethods.GET, '200', {
      one: 12,
      two: 'owt',
    });
    const match = someResponsesWithBodiesFixture.responseDiffsForTraffic(
      traffic,
      exampleMatch
    );

    expect(
      someResponsesWithBodiesFixture.responseContentDiffsForTraffic(
        traffic,
        match.context
      )
    ).toMatchSnapshot();
  });
});
