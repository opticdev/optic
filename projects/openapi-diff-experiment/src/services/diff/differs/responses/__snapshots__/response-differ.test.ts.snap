// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`response differ diff for new status code 1`] = `
Object {
  "context": undefined,
  "diffs": Array [
    Object {
      "method": "get",
      "path": "/example",
      "statusCode": "200",
      "type": "UnmatchedResponse",
    },
  ],
  "error": undefined,
  "isDiff": true,
  "isMatch": false,
}
`;

exports[`response differ json schema diffs found in responses 1`] = `
Object {
  "context": undefined,
  "diffs": Array [
    Object {
      "example": Object {
        "one": 12,
        "two": "owt",
      },
      "instancePath": "/two",
      "key": "two",
      "keyword": "additionalProperties",
      "location": Object {
        "inResponse": Object {
          "body": Object {
            "contentType": "application/json",
          },
          "statusCode": "200",
        },
        "jsonSchemaTrail": Array [],
        "method": "get",
        "path": "/example",
      },
      "parentObjectPath": "/additionalProperties",
      "propertyExamplePath": "/two",
      "propertyPath": "/additionalProperties/two",
      "schemaPath": "/schema",
      "type": "BodyAdditionalProperty",
    },
    Object {
      "example": 12,
      "instancePath": "/one",
      "key": "one",
      "keyword": "type",
      "location": Object {
        "inResponse": Object {
          "body": Object {
            "contentType": "application/json",
          },
          "statusCode": "200",
        },
        "jsonSchemaTrail": Array [
          "one",
        ],
        "method": "get",
        "path": "/example",
      },
      "propertyPath": "/properties/one",
      "schemaPath": "/schema",
      "type": "BodyUnmatchedType",
    },
  ],
  "error": undefined,
  "isDiff": true,
  "isMatch": false,
}
`;

exports[`response differ matches a known response by exact status code 1`] = `
Object {
  "context": Object {
    "contentTypes": Array [],
    "statusCodeMatcher": "200",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`response differ matches a known response by range of status code 1`] = `
Object {
  "context": Object {
    "contentTypes": Array [],
    "statusCodeMatcher": "4xx",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`response differ matches default response status code if not other matches 1`] = `
Object {
  "context": Object {
    "contentTypes": Array [],
    "statusCodeMatcher": "default",
  },
  "diffs": Array [],
  "error": undefined,
  "isDiff": false,
  "isMatch": true,
}
`;

exports[`response differ no diff for 500 range status code 1`] = `
Object {
  "context": undefined,
  "diffs": Array [],
  "error": undefined,
  "isDiff": true,
  "isMatch": false,
}
`;
